lib/devise-security/models/session_traceable.rb

```diff
@@ -0,0 +1,111 @@
+# frozen_string_literal: true
+
+require_relative "#{DEVISE_ORM}/session_history"
+require 'devise-security/hooks/session_traceable'
+
+module Devise
+  module Models
+    # SessionTraceable takes care of session trail in every authentication.
+    # When a session expires after expiring the +token+, the user
+    # will be asked for credentials again, it means, he/she will be redirected
+    # to the sign in page.
+    module SessionTraceable
+      extend ActiveSupport::Concern
+      include Devise::Models::Compatibility
+
+      included do
+        has_many :session_histories, as: :owner, class_name: session_history_class,
+                                     inverse_of: :owner, dependent: :destroy
+      end
+
+      def self.required_fields(_klass)
+        %i[session_history_class session_ip_verification]
+      end
+
+      # Log user session.
+      # @return [true] if session is allowed to log.
+      # @return [false] if session isn't allowed to log when already exceeded to `:max_active_sessions`
+      #   or session is not valid.
+      def log_traceable_session!(options = {})
+        token = options[:token] || generate_traceable_token
```

Comment:
Wondering if we should call this `session_history_token` instead of just `token`, to make it clear that it's not a token for the session but for its history.

Reply:
I update the docs to make it more clearer what the `token` is.